src/core/lib/iomgr/exec_ctx.cc

```diff
@@ -176,6 +183,27 @@ grpc_millis grpc_timespec_to_millis_round_up(gpr_timespec ts) {
   return timespec_to_atm_round_up(ts);
 }
 
+void grpc_exec_ctx_maybe_update_start_time(grpc_exec_ctx* exec_ctx) {
+  grpc_exec_ctx_invalidate_now(exec_ctx);
+  grpc_millis now = grpc_exec_ctx_now(exec_ctx);
+  grpc_millis last_start_time_update = gpr_atm_acq_load(&g_last_start_time_update);
+  if (now > last_start_time_update &&
+      now - last_start_time_update > GRPC_START_TIME_UPDATE_INTERVAL) {
+    gpr_timespec real_now = gpr_now(GPR_CLOCK_REALTIME);
+    gpr_timespec old_now = grpc_millis_to_timespec(now, GPR_CLOCK_REALTIME);
+    gpr_timespec diff = gpr_time_sub(real_now, old_now);
+    if (GRPC_TRACER_ON(grpc_timer_check_trace)) {
+      gpr_log(GPR_DEBUG, "Update start time with diff: %" PRId64 "s %dns", diff.tv_sec, diff.tv_nsec);
+    }
+    gpr_mu_lock(&g_start_time_mu);
+    if (last_start_time_update == gpr_atm_acq_load(&g_last_start_time_update)) {
```

Comment:
What is against just overriding the old time with real_now? Why all the diff calculations

Reply:
Because `next` at [here](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/timer_manager.cc#L196) is relative time from monotonic clock. It is added up with g_start_time[GPR_CLOCK_REALTIME] to determine the time until which `gpr_cv_wait` should wait.

Reply:
Actually I just got your idea, Noah. You're right. I fixed that part.